source/slang/slang-ir-use-uninitialized-values.cpp

```diff
@@ -277,30 +285,47 @@ namespace Slang
         }
     }
 
-    static void cancelLoads(ReachabilityContext &reachability, const List<IRInst*>& stores, List<IRInst*>& loads)
+    static bool canStoreReachLoad(ReachabilityContext& reachability, IRInst* store, IRInst* load)
+    {
+        if (reachability.isInstReachable(store, load))
+            return true;
+
+        // Special cases
+        
+        // Target switches; treat as reachable from any of its cases
```

Comment:
> Target switches are by default not reachable by each of their cases

Would it be possible to add an option/Flag to `ReachabilityContext` such that we don't need to have code that handles an "exception" to the rule of reachability explicitly for `__target_switch`?

Reply:
What if `load` is an inst inside a target switch case? It doesn't seem like this logic will capture that.

Reply:
> Would it be possible to add an option/Flag to ReachabilityContext
Yes, we could, but I'm not sure its absolutely necessary for the core purpose of `ReachabilityContext`. The exception with target switch for reachability is only for the warnings so far.

> What if load is an inst inside a target switch case? It doesn't seem like this logic will capture that.
If there was an instruction explicitly loading from the parameter then it would be found in the `IRUse` chain for it, so it doesn't need to be handled in a special way.

Reply:
From what I read the code is doing here: if `load` is a target switch and we want to check if a `store` can reach this target switch, then we go through all the target switch cases and check if that store is reachable from the case block, if it is reachable from any case block, then we say yes?

I don't quite understand  the logic behind this. We are checking if a store can reach a target switch, but then why do we do so by checking if the case blocks can reach the store? It seems reversed...

Reply:
It is indeed reversed, and that is because in this case the store instructions would appear after the target switch (and its case blocks). If we did `isBlockReachable(upper, caseBlock)` instead, it would always be false in this case because a case block is not reachable from an inner instruction.

Reply:
Then why `canStoreReachLoad` is called with `load = target_switch`? If `load` is before `store`, then the function should always return false as the name suggests.

Reply:
We want to make an exception for `__target_switch`s as we want to include a warning if they are the last instruction of block (and if they don't initialize the out parameter). Just like for `IRReturn`s we treat `__target_switch` as a load instruction so that it goes through `cancelLoadStores`. For it to work out in `cancelLoadStores` we do special handling in `canStoreReachLoad` where for target switches only we allow the store to appear after the load.

I could refactor the method name to something like `canStoreCancelLoad` if it makes more sense.